### ISSUE TITLE
NuttX segfault when connected to USB

### ISSUE THREAD

**Issue by Zefz** [PX4/Firmware#1605]
When connected to USB, the Pixhawk from the 3DR Iris+ drone series segfaults at random occasions. This happens on both Debian Linux and OSX. The firmware on the drone was Stable version. Replacing the Pixhawk with a brand new Pixhawk (not from any drone kit) resolves this issue, which is strange since according to Craig Elder from 3DRobotics the hardware should be identical. This issue has been confirmed on two separate Iris+ Pixhawks.

The Pixhawk does not segfault when not connected to USB.

Boot that crashed:
```
[boot] Initialized SPI port 1 (SENSORS)
[boot] Initialized SPI port 2 (RAMTRON FRAM)
[boot] Initialized SPI port 4
[boot] Initialized SDIO
[init] Looking for microSD... 
[init] microSD mounted: /fs/microsd 
[init] Init script not found: /fs/microsd/etc/rc.txt 
[init] AUTOSTART mode 
sercon: Registering CDC/ACM serial driver
sercon: Successfully registered the CDC/ACM serial driver
nshterm [6:70]
[uorb] ready
mtd: failed to set bus speed
param: selected parameter default file /fs/mtd_params
[param] Loaded: /fs/mtd_params 
[rgbled] on I2C bus 1 at 0x55
nsh: pca8574: command not found
 1:   SYS_USE_IO: match
 10016: + SYS_AUTOSTART: match
[init] Config not found: /fs/microsd/etc/config.txt 
CRCs match
Assertion failed at file:armv7-m/up_hardfault.c line: 184 task: init
sp:     20008844
IRQ stack:
  base: 200088b0
  size: 00000ffc
20008840: 200087ac 20008844 100035c0 00000bb4 080b112b 000000b8 080e2940 200087ac
20008860: 080b1361 080beaf9 080b1201 00000010 080b1405 080b13e9 080c1c49 20005bc4
20008880: 00000003 200087ac 080beb25 00000000 20008bb4 a15414fa 000002e6 080be679
200088a0: 080beaf9 00000000 080b1361 200087ac 00000000 10000010 00000001 00000000
sp:     00000000
User stack:
  base: 100035c0
  size: 00000bb4
R0: 080ab1cd 080e28df 080ab1ed 080e28df 080ab1ed 20008848 080e2d93 080ab1b9
R8: 080e0210 080ab1b9 00000bb4 080bc5c1 080ab1ed 080ab1ed 20008848 080e2d93
xPSR: 080e2d93 BASEPRI: 080ab195 CONTROL: 00000000
EXC_RETURN: 080c3319
```

Boot that did not crash:
```
[boot] Initialized SPI port 1 (SENSORS)
[boot] Initialized SPI port 2 (RAMTRON FRAM)
[boot] Initialized SPI port 4
[boot] Initialized SDIO
[init] Looking for microSD...
[init] microSD mounted: /fs/microsd
[init] Init script not found: /fs/microsd/etc/rc.txt
[init] AUTOSTART mode
sercon: Registering CDC/ACM serial driver
sercon: Successfully registered the CDC/ACM serial driver
nshterm [6:70]
[uorb] ready
mtd: failed to set bus speed
param: selected parameter default file /fs/mtd_params
[param] Loaded: /fs/mtd_params
[rgbled] on I2C bus 1 at 0x55
nsh: pca8574: command not found
 1:   SYS_USE_IO: match
 10016: + SYS_AUTOSTART: match
[init] Config not found: /fs/microsd/etc/config.txt
CRCs match
dataman: Unknown restart
dataman: Initialized, data manager file '/fs/microsd/dataman' size is 101504 bytes
commander: starting
commander: loaded mission state: dataman_id=0, count=0, current=0
[px4io] default PWM output device
[init] PX4IO started
[init] Set PX4IO update rate limit: 400 Hz
mavlink_if0: mode: NORMAL
mavlink_if0: data rate: 1000 Bytes/s, port: /dev/ttyS1, baud: 57600
mavlink_if0: offboard mission init: dataman_id=0, count=0, current_seq=0
[MS5611_SPI] on SPI bus 1 at 3
<adc> init done
mpu6000: driver start failed
[MPU6000] on SPI bus 1 at 4
<L3GD20> probe failed
l3gd20: driver start failed
[L3GD20] on SPI bus 1 at 1
[HMC5883] on I2C bus 1 at 0x1e
ver hwcmp match: PX4FMU_V2
<LSM303D> probe failed
lsm303d: SPI init failed
lsm303d: driver start failed
[LSM303D] on SPI bus 1 at 2
meas_airspeed: no MS4525 airspeed sensor connected
ets_airspeed: no ETS airspeed sensor connected
ets_airspeed: no ETS airspeed sensor connected
px4flow: driver start failed
preflight_check [95:100]
Start sdlog2 at 200Hz
[init] Vehicle type: MULTICOPTER
  MAV_TYPE: curr: 1 -> new: 2
[px4io] mixer sent
[init] Mixer loaded: /etc/mixers/FMU_quad_w.mix
preflight_check: magnetometer calibration missing or bad - calibrate magnetometer first
preflight_check: PREFLIGHT CHECK ERROR! TRIGGERING ALARM
gps: ubx msg 0x0103 len 16 unexpected
gps: ubx disabling msg 0x0103
gps: ubx msg 0x0501 invalid len 25269
navigator: Initializing..
[init] No addons script: /fs/microsd/etc/extras.txt

NuttShell (NSH)
nsh> navigator: Could not clear geofence
navigator: offboard mission updated: dataman_id=0, count=0, current_seq=0
navigator: current offboard mission index: 0
sdlog2: log buffer size: 22528 bytes
mavlink_if0: DISABLING HARDWARE FLOW CONTROL
position_estimator_inav: baro offs: 178
gps: ubx msg 0x0601 NAK
gps: not using NAV-PVT
gps: VER hash 0xe0cc2322
gps: VER hw  "  00040007"
gps: VER sw  "                  7.03 (45970)"
gps: VER ext "                   7.03 (45969)"
gps: module found: UBX
```

Boot that crashed:
```
[boot] Initialized SPI port 2 (RAMTRON FRAM)
[boot] Initialized SPI port 4
[boot] Initialized SDIO
[init] Looking for microSD...
[init] microSD mounted: /fs/microsd
[init] Init script not found: /fs/microsd/etc/rc.txt
[init] AUTOSTART mode
sercon: Registering CDC/ACM serial driver
sercon: Successfully registered the CDC/ACM serial driver
nshterm [6:70]
[uorb] ready
mtd: failed to set bus speed
param: selected parameter default file /fs/mtd_params
[param] Loaded: /fs/mtd_params
[rgbled] on I2C bus 1 at 0x55
nsh: pca8574: command not found
 1:   SYS_USE_IO: match
 10016: + SYS_AUTOSTART: match
[init] Config not found: /fs/microsd/etc/config.txt
CRCs match
Assertion failed at file:armv7-m/up_hardfault.c line: 184 task: init
sp:     20008844
IRQ stack:
  base: 200088b0
  size: 00000ffc
20008840: 200087ac 20008844 100035c0 00000bb4 080b112b 000000b8 080e2940 200087ac
20008860: 080b1361 080beaf9 080b1201 00000010 080b1405 080b13e9 080c1c49 20005bc4
20008880: 00000003 200087ac 080beb25 00000000 20008bb4 215014da 000002e6 080be679
200088a0: 080beaf9 00000000 080b1361 200087ac 00000000 10000010 00000001 00000000
sp:     00000000
User stack:
  base: 100035c0
  size: 00000bb4
R0: 080ab1cd 080e28df 080ab1ed 080e28df 080ab1ed 20008848 080e2d93 080ab1b9
R8: 00000003 080ab1b9 00000bb4 080bc5c1 080ab1ed 080ab1ed 20008848 080e2d93
xPSR: 080e2d93 BASEPRI: 080ab195 CONTROL: 00000000
EXC_RETURN: 080c3319
```

Boot that crashed:
```
[boot] Initialized SPI port 1 (SENSORS)
[boot] Initialized SPI port 2 (RAMTRON FRAM)
[boot] Initialized SPI port 4
[boot] Initialized SDIO
[init] Looking for microSD... 
[init] microSD mounted: /fs/microsd 
[init] Init script not found: /fs/microsd/etc/rc.txt 
[init] AUTOSTART mode 
sercon: Registering CDC/ACM serial driver
sercon: Successfully registered the CDC/ACM serial driver
nshterm [6:70]
[uorb] ready
mtd: failed to set bus speed
param: selected parameter default file /fs/mtd_params
[param] Loaded: /fs/mtd_params 
[rgbled] on I2C bus 1 at 0x55
nsh: pca8574: command not found
 1:   SYS_USE_IO: match
 10016: + SYS_AUTOSTART: match
[init] Config not found: /fs/microsd/etc/config.txt 
CRCs match
Assertion failed at file:armv7-m/up_hardfault.c line: 184 task: init
sp:     20008844
IRQ stack:
  base: 200088b0
  size: 00000ffc
20008840: 200087ac 20008844 100035c0 00000bb4 080b112b 000000b8 080e2940 200087ac
20008860: 080b1361 080beaf9 080b1201 00000010 080b1405 080b13e9 080c1c49 20005bc4
20008880: 00000003 200087ac 080beb25 00000000 20008bb4 255434f2 000002e6 080be679
200088a0: 080beaf9 00000000 080b1361 200087ac 00000000 10000010 00000001 00000000
sp:     00000000
User stack:
  base: 100035c0
  size: 00000bb4
R0: 080ab1cd 080e28df 080ab1ed 080e28df 080ab1ed 20008848 080e2d93 080ab1b9
R8: 00000020 080ab1b9 00000bb4 080bc5c1 080ab1ed 080ab1ed 20008848 080e2d93
xPSR: 080e2d93 BASEPRI: 080ab195 CONTROL: 00000000
EXC_RETURN: 080c3319
```

Boot that crashed:
```
[boot] Initialized SPI port 1 (SENSORS)
[boot] Initialized SPI port 2 (RAMTRON FRAM)
[boot] Initialized SPI port 4
[boot] Initialized SDIO
[init] Looking for microSD... 
[init] microSD mounted: /fs/microsd 
[init] Init script not found: /fs/microsd/etc/rc.txt 
[init] AUTOSTART mode 
sercon: Registering CDC/ACM serial driver
sercon: Successfully registered the CDC/ACM serial driver
nshterm [6:70]
[uorb] ready
mtd: failed to set bus speed
param: selected parameter default file /fs/mtd_params
[param] Loaded: /fs/mtd_params 
[rgbled] on I2C bus 1 at 0x55
nsh: pca8574: command not found
 1:   SYS_USE_IO: match
 10016: + SYS_AUTOSTART: match
[init] Config not found: /fs/microsd/etc/config.txt 
CRCs match
Assertion failed at file:armv7-m/up_hardfault.c line: 184 task: init
sp:     20008844
IRQ stack:
  base: 200088b0
  size: 00000ffc
20008840: 200087ac 20008844 100035c0 00000bb4 080b112b 000000b8 080e2940 200087ac
20008860: 080b1361 080beaf9 080b1201 00000010 080b1405 080b13e9 080c1c49 20005bc4
20008880: 00000003 200087ac 080beb25 00000000 20008bb4 255434f2 000002e6 080be679
200088a0: 080beaf9 00000000 080b1361 200087ac 00000000 10000010 00000001 00000000
sp:     00000000
User stack:
  base: 100035c0
  size: 00000bb4
R0: 080ab1cd 080e28df 080ab1ed 080e28df 080ab1ed 20008848 080e2d93 080ab1b9
R8: 00000020 080ab1b9 00000bb4 080bc5c1 080ab1ed 080ab1ed 20008848 080e2d93
xPSR: 080e2d93 BASEPRI: 080ab195 CONTROL: 00000000
EXC_RETURN: 080c3319
```

**Comment by Zefz** [comment]
I can confirm that the Pixhawk also hangs when connected directly to the Pixhawk instead of through the Iris hull.

**Comment by LorenzMeier** [comment]
I presume master crashes the same way? You can flash via QGC if you want to avoid building.

**Comment by LorenzMeier** [comment]
Would you mind attaching the binary or a drop box link again? I think you already shared it, but I can't seem to find that email, sorry!

**Comment by Zefz** [comment]
Stable Firmware, confirmed segfaults:
https://drive.google.com/file/d/0B7HSl0pSS6x3Sko4bmNhS3NyY1k/view?usp=sharing

**Comment by Zefz** [comment]
It seems master branch has the same behaviour as stable.

**Comment by strykejern** [comment]
Using stable (920cd26) we (Zefz and I) now get this crash report:

```
[boot] Initialized SPI port 1 (SENSORS)
[boot] Initialized SPI port 2 (RAMTRON FRAM)
[boot] Initialized SPI port 4
[boot] Initialized SDIO
[i] Looking for microSD... 
[i] microSD mounted: /fs/microsd 
[i] Init script not found: /fs/microsd/etc/rc.txt 
[i] AUTOSTART mode 
sercon: Registering CDC/ACM serial driver
sercon: Successfully registered the CDC/ACM serial driver
nshterm [6:70]
[uorb] ready
param: selected parameter default file /fs/mtd_params
[param] Loaded: /fs/mtd_params 
[rgbled] on I2C bus 2 at 0x55
nsh: pca8574: command not found
 1:   SYS_USE_IO: match
 10016: + SYS_AUTOSTART: match
[i] Config not found: /fs/microsd/etc/config.txt 
Assertion failed at file:armv7-m/up_hardfault.c line: 184 task: px4io
sp:     200092d4
IRQ stack:
  base: 20009340
  size: 00000ffc
200092c0: 000004ac 080adc01 20009340 000004ac 080adc01 200092d4 10008208 20009340
200092e0: 080adc15 000000b8 080debac 2000923c 080ade45 080bb525 080adce5 00000010
20009300: 080adee9 080adecd 080be631 2000665c 00000003 2000923c 080bb551 00000000
20009320: 200095d4 210908a4 000002e6 080bb12d 080bb525 00000000 080ade45 2000923c
sp:     00000000
User stack:
  base: 10008208
  size: 000004ac
R0: 080a7d85 080deb4b 080a7da5 080deb4b 080a7da5 200092d8 080defff 080a7d71
R8: 080af979 080a7d71 000004ac 080b9049 080a7da5 080a7da5 200092d8 080defff
xPSR: 080defff BASEPRI: 080a7d4d CONTROL: 00000000
EXC_RETURN: 080bfcd1
```

The assertion now fails at the task px4io.

**Comment by Zefz** [comment]
Hmm, I hadn't noticed this before. It seems when the USB is connected to Linux, the startup tone output has hiccups and plays much slower compared to OSX or using just the battery.

Could this be some blocking I/O issue on the USB?

**Comment by LorenzMeier** [comment]
@hauptmech This is the binary for your crash report
https://dl.dropboxusercontent.com/u/87745816/firmware_issue_1605_hauptmech.elf

I would assume this was a NSH issue that we fixed yesterday.

```
Assertion failed at file:armv7-m/up_hardfault.c line: 184 task: nshterm
sp:     20009304
IRQ stack:
  base: 20009370
  size: 00000ffc
20009300: 10004834 20009304 10004a88 20009370 080afae9 000000b8 080e0ee8 10004834
20009320: 080afd19 080bd4c5 080afbb9 00000010 080afdbd 080afda1 080c05d1 20006688
20009340: 00000003 10004834 080bd4f1 00000000 1000bc60 1000bc9f 1000bb80 1000bc60
20009360: 1000bc9c 080a4a21 080afd19 10004834 00000000 10000010 00000001 00000000
sp:     10004908
User stack:
  base: 10004a88
  size: 0000063c
10004900: 00000000 080a28a1 1000bc60 1000bc9c 080a32df 080a4a05 1000499c 1000bc60
10004920: 1000bc9c 080a3b1b 1000499c 080a4a21 080a4a05 1000bc60 080ded5c 00000002
10004940: 080ded3c 080a32df 1000bc60 00000002 00000000 00000000 1000499c 080a29e1
10004960: ffffffff 080a3b1b 00000000 080b8be5 100049c7 080b3043 00000001 ffffffff
10004980: 080a29d5 1000bcb2 1000bc9c 080b2fe7 00000000 080a2e35 00000001 1000bc9c
100049a0: 1000bc9f 00000000 00000000 00000000 00000000 00000000 00000000 00000000
100049c0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
100049e0: 1000bc60 080a25db 1000bc9c 1000419c 080a8755 00000000 00000000 080a29e1
10004a00: 1000bc60 080a25d5 10004924 1000bc60 10004a38 10003e7c 00000003 080a25db
10004a20: 00000004 080a0575 0017322e 00000000 00000100 00000000 00050000 00000000
10004a40: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
10004a60: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 080a8755
10004a80: 00000000 00000000 ffffffff 00000000 200003d0 00000000 00000020 80000650
R0: 1000bc60 080a28a1 00000000 75707475 1000bc60 1000bc9f 1000bb80 1000bc60
R8: 1000bc9c 080a4a21 10004490 1000bc9c 080b8a21 10004908 080a28ad a0000000
xPSR: 21000000 BASEPRI: 00000000 CONTROL: 00000000
EXC_RETURN: ffffffe9
```

**Comment by LorenzMeier** [comment]
@Zefz Could you please flash master from here:
http://px4.oznet.ch/px4/continuous/

And let us know if that fixes things for you?

**Comment by LorenzMeier** [comment]
@strykejern I need the *exact* binary you're flashing, not just link / git hash / branch to be able to make use of the stack dump. When you report a crash, please upload the binary at the same time with a *public link*. I appreciate all the reports you've sent in, but I have a very hard time matching the stack dump to a particular binary, and so the hunt is currently quite inefficient.

**Comment by strykejern** [comment]
@LorenzMeier We (Zefz and I) used the firmware included in your link in the previous post, http://px4.oznet.ch/px4/continuous/px4fmu-v2_default.px4

Do we need to reupload it?

**Comment by LorenzMeier** [comment]
You could try raising this here to 2500 and see if that fixes it.
https://github.com/PX4/Firmware/blob/master/src/drivers/px4io/px4io.cpp#L837

**Comment by Zefz** [comment]
Raising the PX4IO stack size to 2500 did not help.

I'll check tomorrow if unplugging the 3DR Telemetry module in the *Stable* branch also avoids segfaulting (currently this trick was only tested on the land_detector branch).

**Comment by Zefz** [comment]
@LorenzMeier The Iris+ Pixhawk segfaults with all ports disconnected (except Serial 4/5 and USB) and with a fresh SD card. Strangely enough this does *not* happen on a Windows PC! This leads me to believe there is a problem with communicating with Linux via USB. I will try it on a third Linux system to see if I can replicate it.

**Comment by LorenzMeier** [comment]
Could you flash the nuttx_bringup branch (don't forget to update submodules and make archives) and report back? It could well be that you need that update. Should be soonish on master.

**Comment by Zefz** [comment]
Ok, I'll try that. I can confirm that it segfaults via USB connection for 3 different Linux Debian 14.04 systems (Ubuntu and Mint).

**Comment by Zefz** [comment]
Yes! NuttX fixes from #1606  seems to fix the segfault :)

I'll do more extensive tests tomorrow, but it seems to be working fine now.

**Comment by velocoderaptor** [comment]
Any idea on the specific cause/fix for this?
